### PR TITLE
Move developer space content into docs on master branch

### LINKED ISSUE #1659: Dissolve `gh-pages` branch
**Describe the flaw**
Currently developer documentation is outdated and not stored in an easily maintainable place.

**Suggest an improvement**
Documentation that is currently stored in the `gh-pages` branch should be moved to the respective places in the code or corresponding external modules. Outdated information shoud be removed or moved to a maintainable location and updated (e.g. neuron modeling documentation migrated to or reference NESTML).

**Supporting materials**

List of items to be migrated to other parts of the documentation. Assignees check the item / reference a PR when this is done.
Items marked for deletion are checked already, because they will be automatically removed with the branch once all items are ticked.

* [x] [`NEST_Contributor_Agreement.pdf`](https://github.com/nest/nest-simulator/blob/gh-pages/NEST_Contributor_Agreement.pdf) -> how to contribute? user docs  -> #1707
* [x] [`code_review_guidelines.md`](https://github.com/nest/nest-simulator/blob/gh-pages/code_review_guidelines.md)  -> basically remove, like "two reviewers" should go to contributing guide  -> #1707
* [x] [`coding_guidelines_c++.md`](https://github.com/nest/nest-simulator/blob/gh-pages/coding_guidelines_c++.m) -> should be somewhere visible, subpage of getting involved  -> #1707
* [x] [`coding_guidelines_sli.md`](https://github.com/nest/nest-simulator/blob/gh-pages/coding_guidelines_sli.md)  -> remove, irrelevant, still in history
* [x] [`continuous_integration.md`](https://github.com/nest/nest-simulator/blob/gh-pages/continuous_integration.md)   -> <s>better docs in any google search, replace with only nest-specific things like how warnings are handled, etc.</s> covered in #1837
* [x] [`development_workflow.md`](https://github.com/nest/nest-simulator/blob/gh-pages/development_workflow.md) 
  * [x]  -> remove (basic git workflow described in github help pages)  -> #1707
  * [x]  -> make sure *NEST-specific conventions* are mentioned in the "how to contribute" page  -> #1707
* [x] [`eclipse_workflow.md`](https://github.com/nest/nest-simulator/blob/gh-pages/eclipse_workflow.md)   -> some sub-page of getting involved (not top level) -> #1719  
* [x] [`xcode_workflow.md`](https://github.com/nest/nest-simulator/blob/gh-pages/xcode_workflow.md)   -> some sub-page of getting involved (not top level)
* [x] [`extension_modules.md`](https://github.com/nest/nest-simulator/blob/gh-pages/extension_modules.md)   -> nest-extension-module -> #1777 / [nest/nest-extension-module:#1](https://github.com/nest/nest-extension-module/issues/3)
* [x] [`git_migration_plans.md`](https://github.com/nest/nest-simulator/blob/gh-pages/git_migration_plans.md)   -> just wrong. delete
* [x] [`index.md`](https://github.com/nest/nest-simulator/blob/gh-pages/index.md)   -> mostly covered by RtD navigation, dissolve in other docs
* [x] [`model_conversion_3g_4g.md`](https://github.com/nest/nest-simulator/blob/gh-pages/model_conversion_3g_4g.md)   -> remove
* [x] [`model_conversion_5g.md`](https://github.com/nest/nest-simulator/blob/gh-pages/model_conversion_5g.md)   -> remove
* [x] [`multimeter_support.md`](https://github.com/nest/nest-simulator/blob/gh-pages/multimeter_support.md)   -> remove, will come from NESTML automatically
* [x] [`neuron_and_device_models.md`](https://github.com/nest/nest-simulator/blob/gh-pages/neuron_and_device_models.md)   -> nest-extension-module
* [x] [`synapse_models.md`](https://github.com/nest/nest-simulator/blob/gh-pages/synapse_models.md)   -> updated and integrated NESTML & nest-extension-module
* [x] [`neuron_model_naming.md`](https://github.com/nest/nest-simulator/blob/gh-pages/neuron_model_naming.md)   -> refactor, model directory intro **and** NESTML
* [x] [`synapse_model_naming.md`](https://github.com/nest/nest-simulator/blob/gh-pages/synapse_model_naming.md)   -> refactor, model directory intro **and** NESTML
* [x] [`stylesheets`](https://github.com/nest/nest-simulator/blob/gh-pages/stylesheets)   -> nice style, but not used anymore
* [x] [`synapses_overview.md`](https://github.com/nest/nest-simulator/blob/gh-pages/synapses_overview.md)   -> NEST concepts at the beginning of user-docs;  also create a page like this for neurons
* [x] [`tokens_and_datums.md`](https://github.com/nest/nest-simulator/blob/gh-pages/tokens_and_datums.md)   -> sli-dev docs, move to doxygen part of `token` and `datum` classes
* [x] [`variables_parameters_naming.md`](https://github.com/nest/nest-simulator/blob/gh-pages/variables_parameters_naming.md)    -> NESTML documentation

### PR DESCRIPTION
**Pull request by clinssen**
Content on the "developer space" pages lives on a separate branch, which adversely impacts its maintainability (for us developers) and discoverability (for users). Much of the content there is outdated, no longer relevant and/or superseded by content that is already on ReadTheDocs.

The remaining content that seems useful is hereby transferred into the ReadTheDocs pages. The "developer space" can be completely removed after merging this PR, by removing the ``gh-pages`` branch on this repository.

Closes #1659.

### CODE REVIEW

**Comment by jessica-mitchell** [comment]
thanks @clinssen for this PR! 

I think this PR covers the files leftover from #1659, and so that issue could be resolved, correct?

For the extension modules - do we want to keep the file, considering  we have the extension module repository in development? 

Also, just a note that I think further re-arranging of some of these files will be necessary, but I'm happy to have them moved into the documentation!

**Comment by clinssen** [comment]
Thanks for linking #1659! Yes, this PR (repeating the verbiage here so GitHub picks up on it) closes #1659.

Unless @jougs points out otherwise, this is still the main documentation for how to write extension modules (there are no docs other than the readme on https://github.com/nest/nest-extension-module). As you rightfully point out though, these documents will eventually need some more finessing than I've done so far in this PR.

**Review by jessica-mitchell — CHANGES_REQUESTED** [review]
@clinssen looks good! I just have a few minor things that could be fixed up

File: doc/userdoc/contribute/neuron_models.rst [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+   port my_model::check_connection(Node& r, port rp)
+   {
+     SpikeEvent e;
+     e.set_sender(*this);
+     e.set_receiver(r);
+     e.set_rport(rp);
+     return r.connect_sender(e);
+   }
+
+Handling Incoming Events
```
Comment (clinssen) [inline doc/userdoc/contribute/neuron_models.rst:151]:
```suggestion
Handling incoming events
```

**Comment by clinssen** [comment]
Cheers, all suggestions committed!

**Review by jessica-mitchell — APPROVED** [review]
LGTM

**Comment by jougs** [comment]
@clinssen: many thanks for your work on this. May I ask you to check the boxes in #1659 that this PR covers?

If there's only the module documentation left, can you please remove it from the developer space and add it to https://github.com/nest/nest-extension-module by means of a PR? Either as a standalone file, or integrated into the README.md, whichever looks more suitable to you. I will then update the content in the context of #1717.

**Comment by heplesser** [comment]
@jougs In light of you comments I made you reviewer no 2 :).

**Comment by clinssen** [comment]
@jougs: this PR closes that issue and checks all the boxes, except for two (material) changes compared to the earlier suggestions:

* [`xcode_workflow.md`](https://github.com/nest/nest-simulator/blob/gh-pages/xcode_workflow.md)   -> removed
* [`tokens_and_datums.md`](https://github.com/nest/nest-simulator/blob/gh-pages/tokens_and_datums.md)   -> removed

Regarding the move to https://github.com/nest/nest-extension-module, this could be a good test of our intersphinx support. But that escalates the scope of the PR a bit, as I will have to set up the doc build environment and I think this is the first time we'll be deploying intersphinx. (@jessica-mitchell: please tell me if I'm wrong!) If you're OK with it, I'll create a seperate issue.

**Comment by jougs** [comment]
@clinssen: See https://github.com/clinssen/nest-simulator/pull/5 for a PR that takes care of the Token documentation. I already checked off in #1659, anticipating you merge soon.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,11 +1,11 @@
 # Contributing to NEST
 
-The NEST simulator is a scientific tool and as such it is never ready and constantly changing to meet the needs of novel neuroscientific endeavors. Here you find the most important information on how you can contribute to NEST. This document is an excerpt from our [developer space](https://nest.github.io/nest-simulator/), which provides more detailed information.
+The NEST simulator is a scientific tool and as such it is never ready and constantly changing to meet the needs of novel neuroscientific endeavors. Here you find the most important information on how you can contribute to NEST. This document is an excerpt from our [Contributing to NEST documentation](https://nest-simulator.readthedocs.io/en/latest/contribute/index.html), which provides more detailed information.
```
Comment (jougs) [inline CONTRIBUTING.md:3]: Unrelated, but still.

```suggestion
The NEST simulator is a scientific tool and as such it is never finished and constantly changing to meet the needs of novel neuroscientific endeavors. Here you find the most important information on how you can contribute to NEST. This document is an excerpt from our [Contributing to NEST documentation](https://nest-simulator.readthedocs.io/en/latest/contribute/index.html), which provides more detailed information.
```

File: doc/userdoc/contribute/extension_modules.rst [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+1. When you make any change to your module, you must first re-compile and re-install your module.
+1. Then move to the NEST build directory and issue
+
+   .. code-block:: sh
+
+      make -C nest clean
+      make
+      make install
+
+   This rebuilds only the NEST executable.
```
Comment (jougs) [inline doc/userdoc/contribute/extension_modules.rst:155]: This file should be moved to https://github.com/nest/nest-extension-module.

File: doc/userdoc/contribute/neuron_models.rst [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+proxies. This is possible because it sends SpikeEvents, which can be
+communicated globally. Since it has proxies, it cannot be a "local
+receiver", as it does not have an instance on all processes.
+
+So why does having proxies not automatically imply being a local
+receiver? Because some devices are global receivers. Volume
+transmitters receive neuromodulatory spikes from neurons which can be
+anywhere on the physical network, but only send modulatory information
+to synapses that are local to them. Thus they have no proxies but are
+not local receivers.
```
Comment (jougs) [inline doc/userdoc/contribute/neuron_models.rst:221]: This file should be moved to https://github.com/nest/nest-extension-module.

File: doc/userdoc/contribute/synapse_models.rst [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+.. code-block:: C++
+
+   #include "my_synapse.h"
+
+And the following line to the end of the file:
+
+.. code-block:: C++
+
+   register_connection_model< my_synapse >( "my_synapse" );
```
Comment (jougs) [inline doc/userdoc/contribute/synapse_models.rst:106]: This file should be moved to https://github.com/nest/nest-extension-module.

File: doc/userdoc/contribute/synapses_overview.rst [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+actual biological arrival time (time when spike occured at sender plus
+delay) of the spike may be up to max_delay (maximum delay) time units
+in the future. This means, in particular, that Connection objects
+cannot perform any computations that depend on the state of the neuron
+at the time of "biological" spike arrival; they can only use historic
+information.
+
+Another important point is that spikes do NOT pass the Connection
+object in correct order of biological arrival time---they are
+unordered in time.
```
Comment (jougs) [inline doc/userdoc/contribute/synapses_overview.rst:115]: This file should be moved to https://github.com/nest/nest-extension-module.

File: doc/userdoc/contribute/index.rst [hunk 7]
```diff
@@ -47,10 +47,22 @@ If you have a Python example network to contribute, please refer to our
...
+
+Writing neuron models
+---------------------
+
+See :doc:`neuron_models`.
+
+Writing synapse models
+----------------------
+
+See :doc:`synapses_overview` and :doc:`synapse_models`.
```
Comment (jougs) [inline doc/userdoc/contribute/index.rst:65]:
```suggestion
Writing an extension module
---------------------------

See `https://github.com/nest/nest-extension-module`_.

Writing neuron and synapse models
---------------------------------

See `https://nestml.readthedocs.io/en/latest/`_.
```

**Review by jougs — CHANGES_REQUESTED** [review]
Thanks!

I have marked the files that I'd like to see moved to the Extension Module repository, but am otherwise fine (happy even!) with the changes.

**Review by jougs — COMMENTED** [review]
So do I see it right that the only file left now is the XCode workflow documentation?

**Comment by clinssen** [comment]
Yes—left behind, that is... did we have an idea for still fitting this in somewhere?

**Comment by jougs** [comment]
I think the proper place would be https://nest-simulator.readthedocs.io/en/latest/contribute/nest_with_ides.html

**Comment by clinssen** [comment]
Done!

**Review by jougs — APPROVED** [review]
Many thanks for your work on this. I have no more requests for changes and approve wholeheartedly :-)

@jessica-mitchell: can you please quickly indicate if you're still happy even with the changes after your approval?

I assume the next logical step after merging this would be to delete the `gh-pages` branch from the repository?

**Comment by jessica-mitchell** [comment]
Yep everything looks good from my end @jougs

**Comment by jougs** [comment]
And branch deleted. Yay!

**Pull request merged**